src/rgw/rgw_redis_scripts.lua

```diff
@@ -0,0 +1,231 @@
+---
+local function formatResponse(errorCode, errorMessage, data)
+    return {
+        map = {
+            errorCode = errorCode,
+            errorMessage = errorMessage,
+            data = data
+        }
+    }
+end
+
+--- 
+--- Lock functions
+---
+
+--- Assert if the lock is held by the owner of the cookie
+--- @param keys table A single element list - lock name
+--- @param args table A single-element list - cookie 
+--- @return number 0 if the lock is held by the owner of the cookie,
+--- -lerrorCodes.EBUSY if the lock is held by another owner, 
+--- -lerrorCodes.ENOENT if the lock does not exist
+local function assert_lock(keys, args)
+    local name = keys[1]
+    local cookie = args[1]
+    if redis.call('EXISTS', name) == 1 then
+        local existing_cookie = redis.call('GET', name)
+        if existing_cookie == cookie then
+            return formatResponse(0, "", "")
+        else
+            return formatResponse(-lerrorCodes.EBUSY, "Lock is held by another process", "")
+        end
+    end
+    return formatResponse(-lerrorCodes.ENOENT, "Lock does not exist", "")
+end
+
+--- Acquire a lock on a resource.
+--- It sets a key with a cookie value if the key does not exist.
+--- If the key exists and the value is same as cookie, it extends the lock.
+--- If the key exists and the value is different from cookie, it fails.  
+---@param keys table A single element list - lock name
+---@param args table A two-element list - cookie and timeout
+---@return number 0 if the lock is acquired or extended
+local function lock(keys, args)
+    local name = keys[1]
+    local cookie = args[1]
+    local timeout = args[2]
+    local lock_status = assert_lock(keys, args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('PEXPIRE', name, timeout)
+        return formatResponse(0, "", "")
+    elseif lock_status.map.errorCode == -lerrorCodes.ENOENT then
+        redis.call('SET', name, cookie, 'PX', timeout)
+        return formatResponse(0, "", "")
+    end
+    return lock_status
+end
+
+--- Release the lock on a resource.
+--- It deletes the key if the value matches the cookie.
+---@param keys table A single element list - lock name
+---@param args table A single-element list - cookie
+local function unlock(keys, args)
+    local name = keys[1]
+    local cookie = args[1]
+    local lock_status = assert_lock(keys, args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('DEL', name)
+        return formatResponse(0, "", "")
+    end
+    return lock_status
+end
+
+--- Register the functions.
+redis.register_function('lock', lock)
+redis.register_function('unlock', unlock)
+redis.register_function('assert_lock', assert_lock)
+
+-----
+--- Queue functions
+-----
+
+-- Refer to 'Execution under low memory conditions'
+-- in https://redis.io/docs/latest/develop/interact/programmability/eval-intro/
+---
+
+-- Function to generate a string longer than n by concatenating the charset
+local function generateString(n)
```

Comment:
why does the string have to be fro mthe charset?
isn't there a faster way to generate a string of specific size?

Reply:
I couldn't come up for find a better way to do this in Lua. 
An even faster approach would be a quadratic growth/concatenation. i.e double the string every time. 
A tradeoff there is higher margin of error. 

For our requirements, I believe a linear approach is sufficient. 
So, say we need to reserve 1KB it would loop 17 times and a string of length 1054 chars. 

If I double the string every time - we get 5 concatenations but 1984 chars.

Reply:
maybe we can hardcode the reservation size, and just create a string of that size?

Reply:
That can be one of the options.